### PR TITLE
Subscribe txs

### PR DESCRIPTION
**Pull request by kaladinlight**

### CODE REVIEW

**Review by 0xdef1cafe — COMMENTED** [review]
only partially reviewed but comments on what i saw

File: go/coinstacks/cosmos/api/api.go [hunk 1]
```diff
@@ -64,8 +63,9 @@ func New(httpClient *cosmos.HTTPClient, grpcClient *cosmos.GRPCClient, swaggerPa
 		handler: &Handler{
 			httpClient: httpClient,
 			grpcClient: grpcClient,
+			wsClient:   wsClient,
 		},
-		mananger: m,
+		mananger: websocket.NewManager(),
```
[the remark this answers is not in the document]
Comment (kaladinlight) [reply go/coinstacks/cosmos/api/api.go:68]: Haha, good catch

File: go/pkg/cosmos/tx.go [hunk 2]
```diff
@@ -1,12 +1,102 @@
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse totalReceive: %s", res.Result.TotalCount)
+	}
+
+	resTxs = append(resTxs, res.Result.Txs...)
+
+	// filter out duplicate transactions
+	seen := make(map[string]bool)
+	uniqueTxs := []client.TxSearchResponseResultTxs{}
+	for _, t := range resTxs {
```
[the remark this answers is not in the document]
Comment (kaladinlight) [reply go/pkg/cosmos/tx.go:54]: nope, you roll your own everything in golang :P haha

File: go/coinstacks/cosmos/api/api.go [hunk 3]
```diff
@@ -107,9 +107,15 @@ func docsRedirect(w http.ResponseWriter, r *http.Request) {
 	http.Redirect(w, r, "/docs/", http.StatusFound)
 }
 
-func (a *API) Start(errChan chan<- error) {
+func (a *API) Serve(errChan chan<- error) {
 	logger.Info("serving application")
-	go a.mananger.Start()
+
+	if err := a.handler.StartWebsocket(); err != nil {
```
Comment (kaladinlight) [inline go/coinstacks/cosmos/api/api.go:113]: Comments would be useful to know which websocket we are interacting with (server -> cosmos, vs client -> server)

File: go/pkg/websocket/websocket.go [hunk 4]
```diff
@@ -73,30 +76,92 @@ func (c *Connection) Write() {
...
 
-	for {
-		select {
-		case message, ok := <-c.msg:
-			c.conn.SetWriteDeadline(time.Now().Add(writeWait))
-			if !ok {
-				c.conn.WriteMessage(websocket.CloseMessage, []byte{})
-				return
-			}
+	// handle ping from server heartbeat and reset read deadline.
```
Comment (kaladinlight) [inline go/pkg/websocket/websocket.go:79]: Update comment to more accurately describe pong received from client

**Pull request merged**